fix(local activity): updated and expanded the explanation of local activities (#4844)
temporalio/documentation: docs/encyclopedia/activities/local-activity.mdx

slug: /local-activity
toc_max_heading_level: 4
keywords:
  - explanation
  - term
  - timeouts
  - activities
  - local activities
  - workflow
tags:
  - Concepts
  - Activities
  - Durable Execution
---

## What is a Local Activity? {/* #local-activity */}

A Local Activity is an [Activity Execution](/activity-execution) that executes in the same Worker process as the [Workflow Execution](/workflow-execution) that schedules it.

Unlike a regular Activity, a Local Activity never enters an Activity Task Queue. Instead, the Workflow Worker executes it directly in an in-process queue. Because it avoids the round trip through the Temporal Service, a Local Activity has significantly lower latency and produces fewer Event History entries than a regular Activity.

Local Activities help with performance optimization and are not a replacement for regular Activities.

Consider using a Local Activity only when the operation:

- is short-lived (completes in a few seconds), including retries
- can execute in the same binary as the Workflow
- does not require routing to a specific Worker or Task Queue
- does not require global rate limiting

> - is idempotent

Reviewer: Maybe link "idempotent" to this best practice? 
https://docs.temporal.io/develop/python/best-practices/error-handling#make-activities-idempotent

4. The result is returned directly to the Workflow.
5. When the Workflow Task completes, the Worker records a `MarkerRecorded` event containing the Local Activity result.

Unlike regular Activities, scheduling and execution occur entirely within the Worker process. Only the final marker is persisted to Event history.

## Workflow Task heartbeating

Local Activities do **not** support Activity heartbeats. Instead, the SDK supports Workflow Task heartbeating.

Workflow Task heartbeating means that if a Local Activity approaches approximately 80% of the Workflow Task Timeout (10 seconds by default), the Worker completes the current Workflow Task and requests a new one from the Temporal Service. This renews the Worker's authorization to continue executing the Workflow and allows the Local Activity to keep running without exceeding the Workflow Task Timeout.

This enables Local Activities to run longer than a single Workflow Task timeout, but it comes with tradeoffs:

- Each Workflow Task heartbeat adds additional Events to the Event history.
- Signals and other external Workflow events are not processed until the Local Activities finish.

> - Commands generated by the Workflow are not sent to the Temporal Service until either:

Reviewer: ```suggestion
- Commands generated by the Workflow are not sent to the Temporal Service until one of the following occurs:
```

>   - the Local Activity completes, or

Reviewer: ```suggestion
  - the Local Activity completes
```

>   - the next Workflow Task heartbeat occurs.

Reviewer: ```suggestion
  - the next Workflow Task heartbeat occurs
```

> If your operation regularly approaches the Workflow Task timeout, it is usually better implemented as a regular Activity.

Reviewer: Do we have a link for how users can monitor their Local Activities?

Author: We didn't have a link anywhere so I added a little guidance in here directly.

## Failure and durability

Regular Activities are durably tracked by the Temporal Service. Scheduling, completion, retries, and failures are all recorded in the Event history. If a Worker crashes after an Activity completes, the completed Activity is not executed again.

Local Activities behave differently. A Local Activity result becomes durable only when the enclosing Workflow Task successfully completes and records a `MarkerRecorded` event. Until then, execution exists only in Worker memory.

If the Worker crashes before the Workflow Task completes, the Workflow Task is retried, causing Local Activities executed during that Workflow Task to run again.

> Because of this behavior, Local Activities provide at-least-once execution semantics and should always be idempotent.

Reviewer: I was hoping to find a link for "execution semantics," but it appears they're only discussed in the context of Nexus. That's possibly a gap we should look at.

Author: I did find a section in Workers on local Activity shutdown behavior and added it in. We'll probably add more on it later.

Once a `MarkerRecorded` event has been written to the Event history, replay uses the recorded result rather than executing the Local Activity again.

## Mixing Local Activities and regular Activities

A Workflow can freely combine Local Activities and regular Activities.

For example:

> ```text

Reviewer: Using a code block here seems like an odd choice, because it contrasts with the bullet list below. I'd suggest either using a bullet list here, or if you've got more time, consider whether a diagram would illustrate the concept better.

Author: I made this into a little mermaid diagram.

Local Activity A
Local Activity B
Regular Activity C
```

When the Workflow Task completes after scheduling Activity C, the Worker sends commands similar to:

- `MarkerRecorded` (Local Activity A)
- `MarkerRecorded` (Local Activity B)
- `ScheduleActivityTask` (Activity C)

If Activity C later fails or retries, the Workflow replays using the recorded markers for A and B. Those completed Local Activities are not executed again because their results have already been persisted in the Event history. Only Activity C is retried according to its Retry Policy.

The only time completed Local Activities execute again is if the Worker fails before their completion markers are recorded.

> ## Limitations

Reviewer: This section seems redundant to me. You've discussed most of these limitations elsewhere in the article, which isn't all that long. The only one I don't see discussed elsewhere is the need for Timers.
